statistics/builder.go

```diff
@@ -184,3 +187,186 @@ func BuildColumnHist(ctx sessionctx.Context, numBuckets, id int64, collector *Sa
+	if err != nil {
+		return nil, nil, err
+	}
+	hg := NewHistogram(id, ndv, nullCount, 0, tp, numBuckets, collector.TotalSize)
+
+	sampleNum := int64(len(samples))
+	// As we use samples to build the histogram, the bucket number and repeat should multiply a factor.
+	sampleFactor := float64(count) / float64(len(samples))
+
+	// Step1: collect topn from samples
+
+	// the topNList is always sorted by count from more to less
+	topNList := make([]TopNMeta, 0, numTopN)
+	cur, err := codec.EncodeKey(ctx.GetSessionVars().StmtCtx, nil, samples[0].Value)
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	curCnt := float64(0)
+	var corrXYSum float64
+
+	// Iterate through the samples
+	for i := int64(0); i < sampleNum; i++ {
+		corrXYSum += float64(i) * float64(samples[i].Ordinal)
+
+		sampleBytes, err := codec.EncodeKey(ctx.GetSessionVars().StmtCtx, nil, samples[i].Value)
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		// case 1, this value is equal to the last one: current count++
+		if bytes.Equal(cur, sampleBytes) {
+			curCnt += 1
+			continue
+		}
+		// case 2, meet a different value: counting for the "current" is complete
+		// case 2-1, now topn is empty: append the "current" count directly
+		if len(topNList) == 0 {
+			topNList = append(topNList, TopNMeta{Encoded: cur, Count: uint64(curCnt)})
+			cur, curCnt = sampleBytes, 1
+			continue
+		}
+		// case 2-2, now topn is full, and the "current" count is less than the least count in the topn: no need to insert the "current"
+		if len(topNList) >= numTopN && uint64(curCnt) <= topNList[len(topNList)-1].Count {
+			cur, curCnt = sampleBytes, 1
+			continue
+		}
+		// case 2-3, now topn is not full, or the "current" count is larger than the least count in the topn: need to find a slot to insert the "current"
+		j := len(topNList)
+		for ; j > 0; j-- {
+			if uint64(curCnt) < topNList[j-1].Count {
+				break
+			}
+		}
+		topNList = append(topNList, TopNMeta{})
+		copy(topNList[j+1:], topNList[j:])
+		topNList[j] = TopNMeta{Encoded: cur, Count: uint64(curCnt)}
+		if len(topNList) > numTopN {
+			topNList = topNList[:numTopN]
+		}
+		cur, curCnt = sampleBytes, 1
+	}
+
+	// Handle the counting for the last value. Basically equal to the case 2 above.
+	// now topn is empty: append the "current" count directly
+	if len(topNList) == 0 {
+		topNList = append(topNList, TopNMeta{Encoded: cur, Count: uint64(curCnt)})
+	} else if len(topNList) < numTopN || uint64(curCnt) > topNList[len(topNList)-1].Count {
+		// now topn is not full, or the "current" count is larger than the least count in the topn: need to find a slot to insert the "current"
+		j := len(topNList)
+		for ; j > 0; j-- {
+			if uint64(curCnt) < topNList[j-1].Count {
+				break
+			}
+		}
+		topNList = append(topNList, TopNMeta{})
+		copy(topNList[j+1:], topNList[j:])
+		topNList[j] = TopNMeta{Encoded: cur, Count: uint64(curCnt)}
+		if len(topNList) > numTopN {
+			topNList = topNList[:numTopN]
+		}
+	}
```

Comment:
These lines seem duplicated with L250-261. Can we merge them, for example, creating a function for this logic?

Reply:
These tens of lines of code are just like solving an algorithm problem. The logic is coupled tightly. I think it's more confusing to create a separate function.
I don't have a good idea to make it cleaner so I added some comments to try to explain it.